Bug fix: Fix UUIDs truncation for custom_id

anushakolan opened the pull request:
**Why is this change needed?**

This change addresses the bug, https://github.com/langchain-ai/langchain-azure/issues/12.
The issue here is when no `custom_id` is provided by the user while inserting a text or document, the python langchain package generates a UUID and stores it in the DB. This `custom_id` value is being truncated.
This change fixes the bug, so that the `custom_id` is not truncated.

**What is reason for the bug?**

The bug is because, the `custom_id` column is of type `varchar(1000) `. The UUID is of type `UniqueIdentifier`. 

When you use `uuid.uuid()` in SQL Server, it generates a UUID object. If you directly insert this object into a varchar column, SQL Server must implicitly convert the UUID object to a string. Depending on the SQL Server version and the settings, this implicit conversion might not always convert the UUID to its string representation correctly, potentially leading to truncation or other unexpected behavior.

**What is the fix for the bug?**

To fix this bug, the UUID is explicitly casted to str and then stored in the DB. 

**How are the changes tested?**

A new test case is added to verify that the UUIDs for `custom_id` are not truncated. This test does the following.

1. Inserts a single text and its metadata into the database without specifying an ID using the add_texts API.
This allows the python langchain package to generate and use random UUIDs.
2. Uses the returned IDs from the add_texts API to retrieve the inserted document information via the get_by_ids API.
3. Extracts the custom_id value from the document information.
4. Compares the length of the custom_id to the length of a UUID to ensure that the UUID has not been truncated.

Also, all the existing tests are run to ensure these changes does not introduce any new bugs.

Aniruddh25 commented on libs/sqlserver/pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "langchain-sqlserver"
-version = "0.1.1"
+version = "0.1.2"
```
Do we plan to create the new patch after this bug fix is checked in?

anushakolan replied:
Yes, that is the plan

Aniruddh25 commented on libs/sqlserver/langchain_sqlserver/vectorstores.py:
```diff
@@ -1243,9 +1243,9 @@ def _insert_embeddings(
+                        custom_id = str(ids[idx]).upper()
                     else:
-                        ids.append(str(uuid.uuid4()))
+                        ids.append(str(uuid.uuid4()).upper())
```
Looks like it is in python that we generate the UUID if its not provided. See here. uuid.uuid4() 
In description you mention it is the system:

`, the system generates a UUID and stores it in the DB. `

Please be specific - `the python langchain package generates a UUID and stores it in DB`

Aniruddh25 commented on libs/sqlserver/langchain_sqlserver/vectorstores.py:
```diff
@@ -1243,9 +1243,9 @@ def _insert_embeddings(
                     # we generate a uuid and add it to the list of IDs to be returned.
                     if idx < len(ids):
-                        custom_id = ids[idx]
+                        custom_id = str(ids[idx]).upper()
```
Instead of doing conversion of uuid to str, do we really need to keep the datatype of `id` column to be `varchar(1000)` ? Should we instead change the datatype to uuid? 

What are the consequences of doing so?

Aniruddh25 commented on libs/sqlserver/tests/integration_tests/test_vectorstores.py:
```diff
@@ -925,6 +925,30 @@ def test_sqlserver_batch_add_documents_with_batch_size_edited(
+
+    if custom_id is not None:
+        # Verify the length of the custom_id is equal to UUID length
+        assert len(custom_id) == uuid_len
```
can we provide an error message if the test fails?

Aniruddh25 commented on libs/sqlserver/tests/integration_tests/test_vectorstores.py:
```diff
@@ -925,6 +925,30 @@ def test_sqlserver_batch_add_documents_with_batch_size_edited(
+    texts = ["hello"]
+    metadatas = [{"summary": "Good Quality Dog Food"}]
+    uuid_len = 36
+    result = store.add_texts(texts, metadatas)
```
Description says that 

>This allows the database to generate and use random UUIDs.

However, it is NOT the database who generates the random UUIDs, we generate it in python client then send it to the database. Please fix description.

anushakolan replied:
Okay

beccadaniel commented:
>When a UUID is casted to str, it is by default converted to lower case. Since the primary key ID is directly stored as UUID, it is in upper case. To maintain consistency with the primary key, the custom_id is also converted to upper case and stored.

How / where are we doing this conversion to uppercase?

Aniruddh25 commented:
> When a UUID is casted to str, it is by default converted to lower case. Since the primary key ID is directly stored as UUID, it is in upper case. To maintain consistency with the primary key, the `custom_id` is also converted to upper case and stored.

This needs to be removed from description since we are no longer upper casing it

anushakolan commented:
> > When a UUID is casted to str, it is by default converted to lower case. Since the primary key ID is directly stored as UUID, it is in upper case. To maintain consistency with the primary key, the custom_id is also converted to upper case and stored.
> 
> How / where are we doing this conversion to uppercase?

We should not, removed it.
